src/vm/contexts.rs

```diff
@@ -686,19 +686,28 @@ impl <'a,'b> Environment <'a,'b> {
                                         contract_content: &ContractAST, contract_string: &str) -> Result<()> {
         self.global_context.begin();
 
-        runtime_cost!(cost_functions::CONTRACT_STORAGE, self, contract_string.len())?;
-
-        // first, store the contract _content hash_ in the data store.
-        //    this is necessary before creating and accessing metadata fields in the data store,
-        //      --or-- storing any analysis metadata in the data store.
-        self.global_context.database.insert_contract_hash(&contract_identifier, contract_string)?;
-        let memory_use = contract_string.len() as u64;
-        self.add_memory(memory_use)?;
-
-        let result = Contract::initialize_from_ast(contract_identifier.clone(), 
-                                                   contract_content,
-                                                   &mut self.global_context);
-        self.drop_memory(memory_use);
+        // wrap in a closure so that `?` can be caught and the global_context can roll_back()
+        //  before returning.
+        let result = (|| {
+            runtime_cost!(cost_functions::CONTRACT_STORAGE, self, contract_string.len())?;
+
+            if self.global_context.database.has_contract(&contract_identifier) {
+                return Err(CheckErrors::ContractAlreadyExists(contract_identifier.to_string()).into())
+            }
+
+            // first, store the contract _content hash_ in the data store.
+            //    this is necessary before creating and accessing metadata fields in the data store,
+            //      --or-- storing any analysis metadata in the data store.
+            self.global_context.database.insert_contract_hash(&contract_identifier, contract_string)?;
+            let memory_use = contract_string.len() as u64;
+            self.add_memory(memory_use)?;
```

Comment:
Shouldn't the memory accounting be done before the call to `insert_contract_hash`?  Asking because it looks like the contract hash could leak if there's a memory overflow (unless, `self.global_context.database` will get rolled back on the subsequent call to `self.context.roll_back()` below)?

Reply:
It gets rolled back in the subsequent call below (because result is an Err in that case).